### PR TITLE
[ICDS Dashboard ]Exclude school going kids from PSE

### PR DESCRIPTION
**Pull request by Rohit25negi**
Code change adds the aggregation on primary and private forms and adds its effect on the child health monthly table. This new parameter which is added is exclusion criteria i.e all PSE eligible children remain eligible unless he/she starts going primary/private school based on the last form submitted for him.

The new table which is created it had to a distributed table in order to join with `child health monthy`. So added code for that in the migration file. Please see that.
I also added target test case data to validate the change.

the best review commit by commit. 

ticket: https://app.asana.com/0/1112385193248823/1138961090831547

### CODE REVIEW

**Comment by Rohit25negi** [comment]
I have changed the query from the window to using an ordered query to fetch the last form for the case. which I think will be faster than the window one.

File: custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py [hunk 1]
```diff
@@ -30,24 +30,21 @@ def aggregation_query(self):
...
+            person_case_id,
+            admitted_private_school='yes' AS admitted_private_school,
+            date_admission_private_school,
+            returned_private_school='yes' AS returned_private_school,
+            date_return_private_school,
+            admitted_primary_school='yes' AS admitted_primary_school,
+            date_admission_primary_school
+            from "{ucr_tablename}"
+            WHERE state_id = %(state_id)s AND
                 timeend < %(next_month_start)s
```
Comment (calellowitz) [inline custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py]: This query looks at all forms of this type ever submitted. That will be very slow as the number increases even now is likely quite slow. It should only have to look at forms from the current month. If there needs to be carried over data, you can see how other intermediate tables do that by joining with the previous month (like the ccs ones).

**Review by calellowitz — CHANGES_REQUESTED** [review]
In addition to the changes I flagged, are you planning on making an airflow pr so that this will run in the real agg?

File: custom/icds_reports/models/aggregate.py [hunk 2]
```diff
@@ -1470,3 +1473,25 @@ class Meta(object):
 
     _agg_helper_cls = AwwIncentiveAggregationDistributedHelper
     _agg_atomic = False
+
+
+class PrimaryPrivateSchool(models.Model, AggregateMixin):
```
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/models/aggregate.py]: Sure, Will take care the convention

File: custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+          SELECT DISTINCT
+            %(state_id)s AS state_id,
+            LAST_VALUE(supervisor_id) over w as supervisor_id,
+            %(month)s::DATE AS month,
+            LAST_VALUE(person_case_id) over w AS person_case_id,
+            LAST_VALUE(admitted_private_school) over w ='yes' AS admitted_private_school,
+            LAST_VALUE(date_admission_private_school) over w  AS date_admission_private_school,
+            LAST_VALUE(returned_private_school) over w='yes'  AS returned_private_school,
+            LAST_VALUE(date_return_private_school) over w  AS date_return_private_school,
+            CLAST_VALUE(admitted_primary_school) over w='yes'  AS admitted_primary_school,
```
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py]: weird how it passes my local tests and thrown any error.

File: custom/icds_reports/tests/agg_tests/fixtures/primary_private.csv [hunk 4]
```diff
@@ -0,0 +1,4 @@
+doc_id,inserted_at,timeend,state_id,supervisor_id,person_case_id,admitted_private_school,date_admission_private_school,returned_private_school,date_return_private_school,admitted_primary_school,date_admission_primary_school
```
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/tests/agg_tests/fixtures/primary_private.csv]: Yes.

File: custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py (hunk 1 above)
Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py]: Hey, I had thought of that. But there is no past table for this therefore that query will give the wrong result. I can think of two solutions.
1) leave it like this for this release and next month update the query to the full join thing as other places do them.
2) do that change now and after that is deployed and before its aggregated. create the last month table with the current query which will be one time and see all forms.
@calellowitz

**Comment by Rohit25negi** [comment]
> are you planning on making an airflow pr so that this will run in the real agg?

I thought, changing here custom/icds_reports/management/commands/run_aggregation_query.py is sufficient. Where else  I need to make changes

File: custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py (hunk 1 above)
Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py]: @calellowitz  I am looking for your opinion on this.

File: custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rohit25negi) [reply custom/icds_reports/utils/aggregation_helpers/distributed/primary_private_school.py]: THis code id out dates. I changed the query.

File: custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py [hunk 5]
```diff
@@ -352,6 +368,10 @@ def _state_aggregation_query(self, state_id):
               AND df.month = %(start_date)s
               AND child_health.state_id = df.state_id
               AND child_health.supervisor_id = df.supervisor_id
+            LEFT OUTER JOIN "{agg_primary_private_school}" ps ON child_health.mother_id = ps.person_case_id
+              AND ps.month = %(start_date)s
+              AND child_health.state_id = ps.state_id
+              AND child_health.supervisor_id = ps.supervisor_id
```
Comment (emord) [inline custom/icds_reports/utils/aggregation_helpers/distributed/child_health_monthly.py]: how many rows does this return? do we know how adding more joins here will affect performance?

File: custom/icds_reports/utils/migrations.py [hunk 6]
```diff
@@ -14,7 +14,6 @@ def get_view_migrations():
         'agg_ccs_record_monthly.sql',
         'agg_child_health_monthly.sql',
         'daily_attendance.sql',
-        'child_health_monthly.sql',
```
Comment (emord) [inline custom/icds_reports/utils/migrations.py]: why is this being deleted?

**Comment by Rohit25negi** [comment]
This is on hold from program side

**Pull request closed without merging**